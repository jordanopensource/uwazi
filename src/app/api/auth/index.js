import needsAuthorization from './authMiddleware';
import captchaAuthorization from './captchaMiddleware';
import { CaptchaModel } from './CaptchaModel';

export { needsAuthorization, captchaAuthorization, CaptchaModel };
export { comparePasswords, encryptPassword } from './encryptPassword';
export { validatePasswordMiddleWare } from './validatePasswordMiddleWare';
