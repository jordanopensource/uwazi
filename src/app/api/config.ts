import uniqueID from 'shared/uniqueID';
import { Tenant } from './tenants/tenantContext';
import { version } from '../../package.json';

const {
  ROOT_PATH,
  JSON_LOGS,
  UPLOADS_FOLDER,
  CUSTOM_UPLOADS_FOLDER,
  ACTIVITY_LOGS_FOLDER,
  USER_SESSION_SECRET,
  MONGO_URI,
  ELASTICSEARCH_URL,
  DBHOST,
  SENTRY_API_DSN,
  MONGO_CONNECTION_POOL_SIZE,
  ENVIRONMENT,
  FILES_ROOT_PATH,
  QUEUE_NAME,
} = process.env;

const rootPath = ROOT_PATH || `${__dirname}/../../`;
const filesRootPath = FILES_ROOT_PATH || rootPath;

// this needs to be true in order for uwazi to work properly
// when using multiple node processes
const CLUSTER_MODE = process.env.CLUSTER_MODE || false;

const onlyDBHOST = () => (DBHOST ? `mongodb://${DBHOST}/` : 'mongodb://127.0.0.1/');

export const config = {
  VERSION: ENVIRONMENT ? version : `development-${version}`,

  JSON_LOGS: JSON_LOGS || false,

  ENVIRONMENT: ENVIRONMENT || 'development',

  PORT: process.env.PORT || 3000,

  DBHOST: MONGO_URI || onlyDBHOST(),

  mongo_connection_pool_size: Number(MONGO_CONNECTION_POOL_SIZE) || 5,

  rootPath,

  publicAssets: `${rootPath}/public/`,

  userSessionSecret: USER_SESSION_SECRET || uniqueID(),

  elasticsearch_nodes: ELASTICSEARCH_URL ? ELASTICSEARCH_URL.split(',') : ['http://localhost:9200'],

  // db for tenants list and sessions
  SHARED_DB: 'uwazi_shared_db',

  multiTenant: process.env.MULTI_TENANT || false,
  clusterMode: CLUSTER_MODE,
  defaultTenant: <Tenant>{
    name: 'default',
    dbName:
      process.env.DATABASE_NAME ||
      (process.env.NODE_ENV === 'test' ? 'uwazi_testing' : 'uwazi_development'),
    indexName:
      process.env.INDEX_NAME ||
      (process.env.NODE_ENV === 'test' ? 'uwazi_testing' : 'uwazi_development'),
    uploadedDocuments: UPLOADS_FOLDER || `${filesRootPath}/uploaded_documents/`,
    attachments: UPLOADS_FOLDER || `${filesRootPath}/uploaded_documents/`,
    customUploads: CUSTOM_UPLOADS_FOLDER || `${filesRootPath}/custom_uploads/`,
    activityLogs: ACTIVITY_LOGS_FOLDER || `${filesRootPath}/log/`,
    featureFlags: {
      s3Storage: false,
    },
  },
  externalServices: Boolean(process.env.EXTERNAL_SERVICES) || false,

  redis: {
    activated: CLUSTER_MODE,
    host: process.env.REDIS_HOST || 'localhost',
    port: parseInt(process.env.REDIS_PORT || '', 10) || 6379,
  },

  sentry: {
    dsn: SENTRY_API_DSN,
    tracesSampleRate: 0.1,
  },
  s3: {
    endpoint: process.env.S3_ENDPOINT || '',
    bucket: process.env.S3_BUCKET || '',
    credentials: {
      accessKeyId: process.env.S3_ACCESS_KEY_ID || '',
      secretAccessKey: process.env.S3_SECRET_ACCESS_KEY || '',
    },
    batchSize: parseInt(process.env.S3_BATCH_SIZE || '', 10) || 1000,
  },
  githubToken: process.env.GITHUB_TOKEN || '',
  queueName: QUEUE_NAME || 'uwazi_jobs',
};
