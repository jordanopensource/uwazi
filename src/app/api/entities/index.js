import model from './entitiesModel';

export { default } from './entities.js';
export { model };
