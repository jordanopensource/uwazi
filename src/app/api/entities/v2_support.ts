import {
  CreateRelationshipService,
  DeleteRelationshipService,
  DenormalizationService as CreateDenormalizationService,
} from 'api/relationships.v2/services/service_factories';
import { DefaultSettingsDataSource } from 'api/settings.v2/database/data_source_defaults';
import { EntitySchema } from 'shared/types/entityType';
import { TemplateSchema } from 'shared/types/templateType';
import { DefaultRelationshipDataSource } from 'api/relationships.v2/database/data_source_defaults';
import { DefaultTransactionManager } from 'api/common.v2/database/data_source_defaults';
import { TemplateMappers } from 'api/templates.v2/database/TemplateMappers';
import { TemplateDBO } from 'api/templates.v2/database/schemas/TemplateDBO';
import { MatchQueryNode } from 'api/relationships.v2/model/MatchQueryNode';
import { DefaultEntitiesDataSource } from 'api/entities.v2/database/data_source_defaults';
import { EntitiesDataSource } from 'api/entities.v2/contracts/EntitiesDataSource';
import { RelationshipProperty } from 'api/templates.v2/model/RelationshipProperty';
import { arrayBidirectionalDiff } from 'shared/data_utils/arrayBidirectionalDiff';
import { DenormalizationService } from 'api/relationships.v2/services/DenormalizationService';

const newRelationshipsEnabled = async () => {
  const transactionManager = DefaultTransactionManager();
  return DefaultSettingsDataSource(transactionManager).readNewRelationshipsAllowed();
};

const deleteRelatedNewRelationships = async (sharedId: string) => {
  if (await newRelationshipsEnabled()) {
    const datasource = DefaultRelationshipDataSource(DefaultTransactionManager());
    await datasource.deleteByEntities([sharedId]);
  }
};

const withDenormalizationService = async (
  cb: (service: DenormalizationService) => Promise<void>
) => {
  if (await newRelationshipsEnabled()) {
    const transactionManager = DefaultTransactionManager();
    const denormalizationService = await CreateDenormalizationService(transactionManager);
    await cb(denormalizationService);
    await transactionManager.executeOnCommitHandlers(undefined);
  }
};

const denormalizeAfterEntityCreation = async (data: { sharedId: string; language: string }) => {
  await withDenormalizationService(async service =>
    service.denormalizeAfterCreatingEntities([data.sharedId], data.language)
  );
};

const denormalizeAfterEntityUpdate = async (data: { sharedId: string; language: string }) => {
  await withDenormalizationService(async service =>
    service.denormalizeAfterUpdatingEntities([data.sharedId], data.language)
  );
};

const diffMetadataValues = (
  currentDoc: EntitySchema,
  toSave: EntitySchema,
  propertyName: string
) => {
  const diff = arrayBidirectionalDiff(
    currentDoc?.metadata?.[propertyName] ?? [],
    toSave?.metadata?.[propertyName] ?? [],
    v => v.value as string,
    v => v.value as string
  );

  return { newValues: diff.added, deletedValues: diff.removed };
};

interface RelationshipDefinition {
  type: string;
  to: string;
  from: string;
}

interface DefinitionsToUpdate {
  newRelationships: RelationshipDefinition[];
  removedRelationships: RelationshipDefinition[];
}

const determineRelationships = async (
  entitiesDataSource: EntitiesDataSource,
  values: string[],
  entity: EntitySchema,
  query: MatchQueryNode
) => {
  const relationships: RelationshipDefinition[] = [];
  await entitiesDataSource.getByIds(values, entity.language).forEach(async targetEntity => {
    relationships.push(query.determineRelationship(targetEntity));
  });
  return relationships;
};

const ignoreNewRelationshipsMetadata = async (
  currentDoc: EntitySchema,
  toSave: EntitySchema,
  template: TemplateSchema
): Promise<DefinitionsToUpdate> => {
  const newRelationships: RelationshipDefinition[] = [];
  const removedRelationships: RelationshipDefinition[] = [];
  const entitiesDataSource = DefaultEntitiesDataSource(DefaultTransactionManager());
  if (await newRelationshipsEnabled()) {
    const templateModel = TemplateMappers.toApp(template as TemplateDBO);
    await Promise.all(
      templateModel.properties.map(async property => {
        if (property instanceof RelationshipProperty) {
          if (toSave.metadata && currentDoc.metadata) {
            const { newValues, deletedValues } = diffMetadataValues(
              currentDoc,
              toSave,
              property.name
            );

            const query = property.buildQueryRootedInEntity(currentDoc.sharedId!);
            newRelationships.push(
              ...(await determineRelationships(entitiesDataSource, newValues, currentDoc, query))
            );
            removedRelationships.push(
              ...(await determineRelationships(
                entitiesDataSource,
                deletedValues,
                currentDoc,
                query
              ))
            );

            // eslint-disable-next-line no-param-reassign
            toSave.metadata[property.name] = currentDoc.metadata[property.name] || [];
          }
        }
      })
    );
  }
  return { newRelationships, removedRelationships };
};

const createNewRelationships = async (
  relationships: { type: string; to: string; from: string }[]
) => {
  const service = await CreateRelationshipService();
  await service.create(
    relationships.map(r => ({
      type: r.type,
      to: {
        type: 'entity',
        entity: r.to,
      },
      from: {
        type: 'entity',
        entity: r.from,
      },
    }))
  );
};

const deleteRemovedRelationships = async (
  relationships: { type: string; to: string; from: string }[]
) => {
  if (relationships.length) {
    const transactionManager = DefaultTransactionManager();
    const dataSource = DefaultRelationshipDataSource(transactionManager);
    const service = await DeleteRelationshipService();
    const toDelete: string[] = [];

    await dataSource.getByDefinition(relationships).forEach(async relationship => {
      toDelete.push(relationship._id);
    });

    await service.delete(toDelete);
  }
};

const updateNewRelationships = async (updates: DefinitionsToUpdate) => {
  if (await newRelationshipsEnabled()) {
    await createNewRelationships(updates.newRelationships);
    await deleteRemovedRelationships(updates.removedRelationships);
  }
};

export {
  deleteRelatedNewRelationships,
  ignoreNewRelationshipsMetadata,
  updateNewRelationships,
  denormalizeAfterEntityCreation,
  denormalizeAfterEntityUpdate,
};
