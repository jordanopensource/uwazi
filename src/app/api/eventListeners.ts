import { EventsBus } from './eventsbus';
import { AutomaticTranslationFactory } from './externalIntegrations.v2/automaticTranslation/AutomaticTranslationFactory';
import { registerEventListeners as registerSegmentationListeners } from './services/pdfsegmentation/eventListeners';
import { Suggestions } from './suggestions/suggestions';

const registerEventListeners = (eventsBus: EventsBus) => {
  Suggestions.registerEventListeners(eventsBus);
  registerSegmentationListeners(eventsBus);
  AutomaticTranslationFactory.defaultATEntityCreationListener(eventsBus).start();
};

export { registerEventListeners };
