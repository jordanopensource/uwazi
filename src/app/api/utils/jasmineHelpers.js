/** @format */

const assessStatus = routeResult => {
  let status;
  const req = {};
  const next = () => {};
  const res = {
    status: code => {
      status = code;
    },
    json: () => {},
  };

  routeResult.middlewares[0](req, res, next);

  return status;
};

const assessAuthorized = (routeResult, roles = []) => {
  const nextCalledFor = [];

  roles.forEach(role => {
    const req = {
      user: { role },
      get: key => (key === 'X-Requested-With' ? 'XMLHttpRequest' : ''),
    };
    const res = { status: () => {}, json: () => {} };
    const next = () => {
      nextCalledFor.push(role);
    };
    routeResult.middlewares[0](req, res, next);
  });

  return roles.join(',') === nextCalledFor.join(',');
};

const conformValidationErrors = (routeCanFail, expected) =>
  routeCanFail
    ? `Route is not correctly authorized for ['${expected.join("', '")}']`
    : 'Route is not authorized ! (Auth middleware should be the first one)';

const conformMessage = (result, routeCanFail, expected) =>
  result.pass ? 'route is authorized' : conformValidationErrors(routeCanFail, expected);

const matchers = {
  toNeedAuthorization(routeResult, expected) {
    const routeCanFail = assessStatus(routeResult) === 401;
    const routeValidatesExpected = expected ? assessAuthorized(routeResult, expected) : true;
    const result = { pass: routeCanFail && routeValidatesExpected };
    result.message = () => conformMessage(result, routeCanFail, expected);
    return result;
  },
};

expect.extend(matchers);

export function catchErrors(done) {
  return error => {
    if (error instanceof Error) {
      return done.fail(error);
    }
    return done.fail(JSON.stringify(error));
  };
}
