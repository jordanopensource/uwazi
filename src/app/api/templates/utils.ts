import uuid from 'node-uuid';
import { ObjectId } from 'mongodb';
import { differenceBy, intersectionBy } from 'lodash';

import settings from 'api/settings/settings';
import { files } from 'api/files';
import propertiesHelper from 'shared/commonProperties';
import { objectIndex } from 'shared/data_utils/objectIndex';
import { safeName as sharedSafeName } from 'shared/propertyNames';
import { ensure } from 'shared/tsUtils';
import { ExtractedMetadataSchema, PropertySchema } from 'shared/types/commonTypes';
import { TemplateSchema } from 'shared/types/templateType';
import { ThesaurusSchema, ThesaurusValueSchema } from 'shared/types/thesaurusType';
import model from './templatesModel';

const safeName = sharedSafeName;

const getInheritedProps = async (templates: TemplateSchema[]) => {
  const properties: PropertySchema[] = propertiesHelper
    .allUniqueProperties(templates)
    .filter((p: PropertySchema) => p.inherit?.property);

  return (
    await model.db.aggregate([
      {
        $match: {
          'properties._id': {
            $in: properties.map(p => new ObjectId(p.inherit?.property)),
          },
        },
      },
      {
        $project: {
          properties: {
            $filter: {
              input: '$properties',
              as: 'property',
              cond: {
                $or: properties.map(p => ({
                  $eq: ['$$property._id', new ObjectId(p.inherit?.property)],
                })),
              },
            },
          },
          _id: 0,
        },
      },
      { $unwind: '$properties' },
      { $replaceRoot: { newRoot: '$properties' } },
    ])
  ).reduce((indexed, prop) => ({ ...indexed, [prop._id.toString()]: prop }), {});
};

const denormalizeInheritedProperties = async (template: TemplateSchema) => {
  if (template.synced) return template.properties;

  const inheritedProperties: { [k: string]: PropertySchema } = await getInheritedProps([template]);

  return template.properties?.map(prop => {
    if (!prop.inherit?.property) {
      // eslint-disable-next-line no-param-reassign
      delete prop.inherit;
      return prop;
    }

    const { type } = inheritedProperties[prop.inherit.property];
    // eslint-disable-next-line no-param-reassign
    prop.inherit.type = type;
    return prop;
  });
};

const generateName = (property: PropertySchema, newNameGeneration: boolean) => {
  const name = property.label ? safeName(property.label, newNameGeneration) : property.name;
  return property.type === 'geolocation' || property.type === 'nested'
    ? `${name}_${property.type}`
    : name;
};

const generateNames = async (properties: PropertySchema[]) => {
  const { newNameGeneration = false } = await settings.get();
  return properties.map(property => ({
    ...property,
    name: generateName(property, newNameGeneration),
  }));
};

const newThesauriId = () => uuid.v4();

function generateIds(properties: ThesaurusSchema[] = []) {
  return properties.map(property => ({
    ...property,
    id: property.id || newThesauriId(),
  }));
}
interface PropertyOrThesaurusSchema
  extends Partial<PropertySchema>,
    Partial<ThesaurusValueSchema> {}

const flattenProperties = (properties: PropertyOrThesaurusSchema[]) =>
  properties.reduce<PropertyOrThesaurusSchema[]>((flatProps, p) => {
    if (p.values) {
      return [...flatProps, ...p.values, p];
    }

    return [...flatProps, p];
  }, []);

function getUpdatedIds(
  {
    prop,
    filterBy,
  }: {
    prop: 'name' | 'label';
    filterBy: 'id' | '_id';
  },
  oldProperties: PropertyOrThesaurusSchema[] = [],
  newProperties: PropertyOrThesaurusSchema[] = []
) {
  const indexOf = (p: PropertyOrThesaurusSchema) => p[filterBy]!.toString();
  const idsWithNewName: { [k: string]: string } = {};
  const flatOld = flattenProperties(oldProperties);
  const flatNew = flattenProperties(newProperties);
  const newByIndex = objectIndex(flatNew, indexOf, p => p);
  flatOld.forEach(property => {
    const newProperty = newByIndex[indexOf(property)];
    const oldValue = property[prop];
    const newValue = newProperty?.[prop];
    if (newProperty && typeof newValue === 'string' && newValue !== oldValue) {
      idsWithNewName[indexOf(property)] = newValue;
    }
  });
  return idsWithNewName;
}

function getUpdatedNames(
  {
    prop,
    filterBy,
  }: {
    prop: 'name' | 'label';
    filterBy: 'id' | '_id';
  },
  oldProperties: PropertyOrThesaurusSchema[] = [],
  newProperties: PropertyOrThesaurusSchema[] = []
) {
  const indexOf = (p: PropertyOrThesaurusSchema) => p[filterBy]?.toString() || '';

  const propertiesWithNewName: { [k: string]: string } = {};
  const deletedProperties: Set<string> = new Set();

  const flatOld = flattenProperties(oldProperties);
  const preExistingLabels = new Set(flatOld.map(property => property[prop]));
  const flatNew = flattenProperties(newProperties);
  const newByIndex = objectIndex(flatNew, indexOf, p => p);
  const allNewLabels = new Set(flatNew.map(property => property[prop]));

  const labelAlreadyExisted = (p: PropertyOrThesaurusSchema) => preExistingLabels.has(p[prop]);
  const labelStillInUse = (p: PropertyOrThesaurusSchema) => allNewLabels.has(p[prop]);
  const sameProp = (previous: PropertyOrThesaurusSchema, current: PropertyOrThesaurusSchema) =>
    previous[prop] === current[prop];
  const compareProps = (
    previous: PropertyOrThesaurusSchema,
    current: PropertyOrThesaurusSchema
  ) => {
    if (!current || sameProp(previous, current) || labelAlreadyExisted(current)) {
      if (!labelStillInUse(previous)) deletedProperties.add(previous[prop]!);
      return { key: undefined, value: undefined };
    }
    if (labelStillInUse(previous)) {
      return { key: current[prop], value: current[prop] };
    }
    return { key: previous[prop], value: current[prop] };
  };

  flatOld.forEach(property => {
    const newProperty = newByIndex[indexOf(property)];
    const { key, value } = compareProps(property, newProperty);
    if (key && value) {
      propertiesWithNewName[key] = value;
    }
  });

  return { update: propertiesWithNewName, delete: Array.from(deletedProperties) };
}

function getDeletedProperties(
  oldProperties: PropertyOrThesaurusSchema[] = [],
  newProperties: PropertyOrThesaurusSchema[] = [],
  filterBy: 'id' | '_id' = '_id',
  prop: 'name' | 'label' | 'id' = 'name'
) {
  const flatOld = flattenProperties(oldProperties);
  const flatNew = flattenProperties(newProperties);
  const newIds = new Set(flatNew.map(property => property[filterBy]?.toString()));
  const newLabels = new Set(flatNew.map(property => property[prop]?.toString()));
  const filteredById = flatOld
    .filter(
      property => !newIds.has(property[filterBy]?.toString()) && !newLabels.has(property[prop])
    )
    .filter((property): property is { [k: string]: string } => typeof property[prop] === 'string')
    .map(property => property[prop]);
  return filteredById;
}

function getRenamedTitle(
  oldCommonProperties: PropertySchema[],
  newCommonProperties: PropertySchema[]
) {
  const oldTitle = ensure<PropertySchema>(oldCommonProperties.find(p => p.name === 'title'));
  const newTitle = ensure<PropertySchema>(newCommonProperties.find(p => p.name === 'title'));
  return oldTitle.label !== newTitle.label ? [oldTitle.label] : [];
}

const propertyUpdater = async (
  modifiedProperties: PropertySchema[] = [],
  updateFunction: (
    array: ExtractedMetadataSchema[],
    property: PropertySchema
  ) => ExtractedMetadataSchema[] = () => []
) =>
  modifiedProperties.reduce(async (previousPromise: Promise<void>, property) => {
    await previousPromise;
    const affectedFiles = await files.get({
      'extractedMetadata.propertyID': property._id?.toString(),
    });

    await affectedFiles.reduce(async (prevPromise: Promise<void>, file) => {
      await prevPromise;
      await files.save({
        ...file,
        extractedMetadata: updateFunction(file.extractedMetadata || [], property),
      });
    }, Promise.resolve());
  }, Promise.resolve());

const updateExtractedMetadataProperties = async (
  oldProperties: PropertySchema[] = [],
  newProperties: PropertySchema[] = []
) => {
  const currentProperties = oldProperties.map(property => ({
    ...property,
    _id: property._id?.toString(),
  }));

  const differentProperties = differenceBy(newProperties, currentProperties, 'name').filter(
    property => ['text', 'markdown', 'numeric', 'date'].includes(property.type)
  );

  const renamedProperties = intersectionBy(differentProperties, currentProperties, '_id');

  const removedProperties = differenceBy(currentProperties, newProperties, '_id').filter(property =>
    ['text', 'markdown', 'numeric', 'date'].includes(property.type)
  );

  if (removedProperties.length > 0) {
    await propertyUpdater(removedProperties, (metadata, property) =>
      metadata.filter(data => data.propertyID !== property._id?.toString())
    );
  }

  if (renamedProperties.length > 0) {
    await propertyUpdater(renamedProperties, (metadata, property) =>
      metadata.map(data => {
        if (data.propertyID === property._id?.toString()) {
          return { ...data, name: property.name };
        }
        return data;
      })
    );
  }

  return null;
};

export type { PropertyOrThesaurusSchema };
export {
  newThesauriId,
  safeName,
  denormalizeInheritedProperties,
  flattenProperties,
  generateIds,
  getUpdatedIds,
  getUpdatedNames,
  generateNames,
  getDeletedProperties,
  getRenamedTitle,
  updateExtractedMetadataProperties,
};
