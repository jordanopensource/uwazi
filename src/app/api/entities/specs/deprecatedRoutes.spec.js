import { search } from 'api/search';
import 'api/utils/jasmineHelpers';
import db from 'api/utils/testing_db';
import documentRoutes from '../routes.js';
import instrumentRoutes from '../../utils/instrumentRoutes';
import entities from '../entities';
import * as entitiesSavingManager from '../entitySavingManager';
import templates from '../../templates/templates';
import thesauri from '../../thesauri';
import fixtures, { templateId, unpublishedDocId, batmanFinishesId } from './fixtures.js';

describe('entities', () => {
  let routes;

  beforeEach(async () => {
    routes = instrumentRoutes(documentRoutes);
    jest
      .spyOn(search, 'countPerTemplate')
      .mockImplementation(async () => Promise.resolve({ templateCount: 0 }));
    await db.setupFixturesAndContext(fixtures);
  });

  afterAll(async () => {
    await db.disconnect();
  });

  describe('POST', () => {
    let req;
    beforeEach(() => {
      req = {
        body: {
          title: 'Batman begins',
          template: templateId,
        },
        user: { username: 'admin' },
        language: 'lang',
        io: {
          sockets: {
            emit: () => {},
          },
        },
      };
    });

    it('should need authorization', () => {
      expect(routes._post('/api/entities', req)).toNeedAuthorization();
      expect(routes._post('/api/entity_denormalize', req)).toNeedAuthorization();
    });

    it('should create a new document with current user', done => {
      jest
        .spyOn(entitiesSavingManager, 'saveEntity')
        .mockReturnValue(Promise.resolve({ entity: 'entity', errors: [] }));
      jest
        .spyOn(templates, 'getById')
        .mockImplementation(async () => Promise.resolve({ values: [] }));
      jest
        .spyOn(thesauri, 'templateToThesauri')
        .mockImplementation(async () => Promise.resolve('document'));
      const sockets = {
        emitToCurrentTenant: jest.fn(),
      };
      routes.post('/api/entities', { ...req, sockets }).then(document => {
        expect(document).toBe('entity');
        expect(entitiesSavingManager.saveEntity).toHaveBeenCalledWith(req.body, {
          user: req.user,
          language: 'lang',
        });
        done();
      });
    });

    it('should denormalize an entity', async () => {
      jest.spyOn(entities, 'denormalize').mockImplementation(async () => Promise.resolve('entity'));

      const document = await routes.post('/api/entity_denormalize', req);
      expect(document).toBe('entity');
      expect(entities.denormalize).toHaveBeenCalledWith(req.body, {
        user: req.user,
        language: 'lang',
      });
    });

    it('should emit thesauriChange socket event with the modified thesaurus based on the entity template', async () => {
      const user = {
        _id: 'c08ef2532f0bd008ac5174b45e033c93',
        username: 'admin',
      };
      req = {
        body: {
          title: 'Batman begins',
          template: 'template',
        },
        user,
        language: 'lang',
        sockets: {
          emitToCurrentTenant: jest.fn((event, thesaurus) => {
            expect(event).toBe('thesauriChange');
            expect(thesaurus).toBe('templateTransformed');
            expect(thesauri.templateToThesauri).toHaveBeenCalledWith('template', 'lang', user, {
              templateCount: 0,
            });
          }),
        },
      };

      jest
        .spyOn(entitiesSavingManager, 'saveEntity')
        .mockReturnValue(Promise.resolve({ entity: { _id: 'id' }, errors: [] }));
      jest
        .spyOn(entities, 'getWithRelationships')
        .mockReturnValue(Promise.resolve(['entityWithRelationShips']));
      jest.spyOn(templates, 'getById').mockImplementation(async () => Promise.resolve('template'));
      jest
        .spyOn(thesauri, 'templateToThesauri')
        .mockImplementation(async () => Promise.resolve('templateTransformed'));
      await routes.post('/api/entities', req);
    });

    describe('/entities/multipleupdate', () => {
      beforeEach(() => {
        req = {
          body: {
            ids: ['1', '2'],
            values: { metadata: { text: [{ value: 'new text' }] } },
          },
          user: {
            _id: 'c08ef2532f0bd008ac5174b45e033c93',
            username: 'admin',
          },
          language: 'lang',
        };
      });

      it('should need authorization', () => {
        expect(routes._post('/api/entities/multipleupdate', req)).toNeedAuthorization();
      });

      it('should call multipleUpdate with the ids and the metadata in the body', async () => {
        const mockedResponse = [{ sharedId: '1' }, { sharedId: '2' }];
        jest
          .spyOn(entities, 'multipleUpdate')
          .mockImplementation(async () => Promise.resolve(mockedResponse));
        const response = await routes.post('/api/entities/multipleupdate', req);
        expect(entities.multipleUpdate).toHaveBeenCalledWith(
          ['1', '2'],
          { metadata: { text: [{ value: 'new text' }] } },
          {
            user: {
              _id: 'c08ef2532f0bd008ac5174b45e033c93',
              username: 'admin',
            },
            language: 'lang',
          }
        );
        expect(response).toBe(mockedResponse);
      });
    });
  });

  describe('GET', () => {
    it('should return matching document', async () => {
      const expectedEntity = [
        {
          sharedId: 'sharedId',
          published: true,
        },
      ];
      jest
        .spyOn(entities, 'getWithRelationships')
        .mockImplementation(async () => Promise.resolve(expectedEntity));
      const req = {
        query: { sharedId: 'sharedId' },
        language: 'lang',
      };

      const response = await routes.get('/api/entities', req);
      expect(entities.getWithRelationships).toHaveBeenCalledWith(
        {
          sharedId: 'sharedId',
          language: 'lang',
          published: true,
        },
        '',
        { limit: 1 }
      );
      expect(response).toEqual({ rows: expectedEntity });
    });

    it('should return document by id', async () => {
      const expectedEntity = fixtures.entities[0];
      delete expectedEntity.fullText;

      const response = await routes.get('/api/entities', {
        query: { _id: batmanFinishesId, omitRelationships: true },
      });

      expect(response.rows.length).toEqual(1);
      expect(response.rows[0]).toEqual(expect.objectContaining(expectedEntity));
    });

    it('should not return unpublished documents when user not logged in', async () => {
      const response = await routes.get('/api/entities', {
        query: { _id: unpublishedDocId, omitRelationships: true },
      });

      expect(response.status).toEqual(404);
      expect(response.rows).toEqual([]);
    });

    it('should allow not fetching the relationships', async () => {
      const expectedEntity = {
        sharedId: 'sharedId',
        published: true,
      };

      entities.getWithRelationships.mockClear();
      jest.spyOn(entities, 'getWithRelationships').mockImplementation(() => {});
      jest.spyOn(entities, 'get').mockImplementation(async () => Promise.resolve([expectedEntity]));

      const req = {
        query: {
          sharedId: 'sharedId',
          omitRelationships: true,
        },
        language: 'lang',
      };

      const {
        rows: [result],
      } = await routes.get('/api/entities', req);
      expect(result).toBe(expectedEntity);
      expect(entities.getWithRelationships).not.toHaveBeenCalled();
      expect(entities.get).toHaveBeenCalledWith(
        {
          sharedId: 'sharedId',
          language: 'lang',
          published: true,
        },
        '',
        { limit: 1 }
      );
    });

    describe('when the document does not exist', () => {
      it('should retunr a 404', async () => {
        jest
          .spyOn(entities, 'getWithRelationships')
          .mockImplementation(async () => Promise.resolve([]));
        const req = {
          query: { sharedId: 'idontexist' },
          language: 'en',
        };

        const response = await routes.get('/api/entities', req);
        expect(response.status).toBe(404);
      });
    });

    describe('when the document has unpublished relationships and the user is unauthenticated', () => {
      it('should remove private relationships from response', async () => {
        jest.spyOn(entities, 'getWithRelationships').mockReturnValue(
          Promise.resolve([
            {
              sharedId: 'e1',
              published: true,
              relationships: [
                {
                  entityData: {
                    sharedId: 'e1',
                    published: true,
                  },
                },
                {
                  entityData: {
                    sharedId: 'e2',
                    published: false,
                  },
                },
                {
                  entityData: {
                    sharedId: 'e3',
                    published: true,
                  },
                },
              ],
            },
          ])
        );

        const req = {
          query: { sharedId: 'sharedId' },
          language: 'lang',
        };

        const {
          rows: [result],
        } = await routes.get('/api/entities', req);
        expect(result).toEqual({
          sharedId: 'e1',
          published: true,
          relationships: [
            {
              entityData: {
                sharedId: 'e1',
                published: true,
              },
            },
            {
              entityData: {
                sharedId: 'e3',
                published: true,
              },
            },
          ],
        });
      });
    });
  });

  describe('/api/entities/get_raw_page', () => {
    it('should return formattedPlainTextPages page requested', async () => {
      jest.spyOn(entities, 'countByTemplate').mockImplementation(async () => Promise.resolve(2));
      const req = { query: { templateId: 'templateId' } };

      const response = await routes.get('/api/entities/count_by_template', req);

      expect(entities.countByTemplate).toHaveBeenCalledWith('templateId');
      expect(response).toEqual(2);
    });
  });

  describe('/api/entities/count_by_template', () => {
    it('should have a validation schema', () => {
      expect(routes.get.validation('/api/entities/count_by_template')).toMatchSnapshot();
    });
    it('should return count of entities using a specific template', async () => {
      jest.spyOn(entities, 'countByTemplate').mockImplementation(async () => Promise.resolve(2));
      const req = { query: { templateId: 'templateId' } };

      const response = await routes.get('/api/entities/count_by_template', req);
      expect(entities.countByTemplate).toHaveBeenCalledWith('templateId');
      expect(response).toEqual(2);
    });
  });

  describe('DELETE /api/entities', () => {
    beforeEach(() => {
      jest
        .spyOn(entities, 'delete')
        .mockImplementation(async () => Promise.resolve({ json: 'ok' }));
    });

    it('should have a validation schema', () => {
      expect(routes.delete.validation('/api/entities')).toMatchSnapshot();
    });

    it('should use entities to delete it', async () => {
      const req = {
        query: {
          sharedId: 123,
          _rev: 456,
        },
      };
      await routes.delete('/api/entities', req);
      expect(entities.delete).toHaveBeenCalledWith(req.query.sharedId);
    });
  });

  describe('POST /api/entities/bulkdelete', () => {
    beforeEach(() => {
      jest
        .spyOn(entities, 'deleteMultiple')
        .mockImplementation(async () => Promise.resolve({ json: 'ok' }));
    });

    it('should have a validation schema', () => {
      expect(routes.post.validation('/api/entities/bulkdelete')).toMatchSnapshot();
    });

    it('should use entities to delete it', async () => {
      const req = { body: { sharedIds: [123, 456] } };
      await routes.post('/api/entities/bulkdelete', req);
      expect(entities.deleteMultiple).toHaveBeenCalledWith([123, 456]);
    });
  });
});
