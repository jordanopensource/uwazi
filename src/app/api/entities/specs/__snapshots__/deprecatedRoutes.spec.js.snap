// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entities /api/entities/count_by_template should have a validation schema 1`] = `
Object {
  "properties": Object {
    "query": Object {
      "properties": Object {
        "templateId": Object {
          "type": "string",
        },
      },
      "required": Array [
        "templateId",
      ],
      "type": "object",
    },
  },
  "required": Array [
    "query",
  ],
  "type": "object",
}
`;

exports[`entities DELETE /api/entities should have a validation schema 1`] = `
Object {
  "properties": Object {
    "query": Object {
      "properties": Object {
        "sharedId": Object {
          "type": "string",
        },
      },
      "required": Array [
        "sharedId",
      ],
      "type": "object",
    },
  },
  "required": Array [
    "query",
  ],
  "type": "object",
}
`;

exports[`entities POST /api/entities/bulkdelete should have a validation schema 1`] = `
Object {
  "properties": Object {
    "body": Object {
      "properties": Object {
        "sharedIds": Object {
          "items": Object {
            "type": "string",
          },
          "type": "array",
        },
      },
      "required": Array [
        "sharedIds",
      ],
      "type": "object",
    },
  },
  "required": Array [
    "body",
  ],
  "type": "object",
}
`;
